### PR TITLE
Feat/active users metrics

### PR DESCRIPTION
**Pull request by freesteph**
# Description
Add a class to generate active users metrics and upload the result to S3. It delegates the hard work to the code in `PerformancePlatform::Gateway::ActiveUsers` except without the severe indirection of that presenter/gateway/usecase pattern.

This class should be used in a Rake task like this:

```ruby
metrics = Metrics::ActiveUsers.new(period: "week", date: Date.today.to_s)

metrics.generate!

metrics.publish! 
```

The two-step invocation allows the caller to investigate metrics generation and publishing separately, should either of these need to be debugged/altered.

This is the test suite:
```
Metrics::ActiveUsers
  stores the date
  stores the period
  rejects invalid periods
  generate!
    delegates to the performance platform gateway
    stores the result in an instance variable
  key
    returns a key for the S3 upload
  publish!
    when the report is nil
      throws an error
    when the report has been generated
      writes the report into the S3 metrics bucket
```

# How to test
- read commit log;
- checkout locally;
- run `docker-compose exec app bundle exec rake publish_weekly_metrics` and observe output.

### CODE REVIEW

File: spec/lib/metrics/active_users_spec.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  before do
+    @client_double = double
+
+    allow(@client_double).to receive(:put_object)
+    allow(Aws::S3::Client).to receive(:new).and_return(@client_double)
+  end
+
+  it 'stores the date' do
+    expect(subject.period).to eq 'month'
```
Comment (robinmitra) [inline spec/lib/metrics/active_users_spec.rb:14]:
```suggestion
	expect(subject.date).to eq Date.today.to_s
```

File: spec/lib/metrics/active_users_spec.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  it 'stores the period' do
+    expect(subject.date).to eq Date.today.to_s
```
Comment (robinmitra) [inline spec/lib/metrics/active_users_spec.rb:18]:
```suggestion
    expect(subject.period).to eq 'month'
```

File: lib/metrics/active_users.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      gateway = PerformancePlatform::Gateway::ActiveUsers.new(
+        period: period,
+        date: date
+      )
+
+      @report = gateway.fetch_stats
+    end
+
+    def key
+      "logging-api-active-users-#{period}-#{date}"
```
Comment (robinmitra) [inline lib/metrics/active_users.rb:33]: My preference is this, in support of keeping the data decoupled from the code which generates it (as the code may very well move to another place in future, and the data isn't linked to logging API):
```suggestion
      "active-users-#{period}-#{date}"
```

File: spec/lib/metrics/active_users_spec.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+describe Metrics::ActiveUsers do
+  subject { Metrics::ActiveUsers.new(period: 'month', date: Date.today.to_s) }
+
+  before do
+    @client_double = double
```
Comment (robinmitra) [inline spec/lib/metrics/active_users_spec.rb:7]: My suggestion is to rename this to `@s3_client_double` to make it explicit.

**Comment by alan-gds** [comment]
Oops.  I rebased my linting updates into this branch, but it looks like some work needs to be done to apply the new linting standards to the work that's been done here so far.  I'll sort this out and commit a fix (doing it as a rebase would probably be unsafely fiddly).

File: Gemfile.lock [hunk 5]
```diff
@@ -138,13 +172,14 @@ PLATFORMS
...
   aws-sdk-s3 (~> 1)
+  guard-rspec
   mysql2
   puma
   rack-test
   rake (~> 13.0)
   require_all
   rspec
-  rubocop-govuk (~> 3.16.0)
+  rubocop-govuk
```
Comment (alan-gds) [inline Gemfile.lock:182]: This one's my fault.  The version number should not have been pinned in the first place.  It's a mistake I introduced in the master branch.  Bundler has spotted it and removed the redundant pin; we might as well let the correction be made in this PR.

**Comment by alan-gds** [comment]
I've committed fixes for the linting issues.  As per the commit message, I recommend someone reviews what I changed and then squashes it into the appropriate earlier commit (I'm fairly sure it's fine but I don't want to be too gung-ho with force-pushed rebasing).

File: spec/lib/metrics/active_users_spec.rb (hunk 2 above)
Comment (freesteph) [reply spec/lib/metrics/active_users_spec.rb:18]: that's right above

**Comment by freesteph** [comment]
@robinmitra I've amended the code according to your comments. I've committed "fixup!" commits for you to see instead of straight rebasing-force-pushing. Once you're happy with it let me know and I can push a final, clean, rebased branch.

**Pull request merged**